modules/python/python-logtemplate.c

```diff
@@ -0,0 +1,157 @@
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+#include "python-logtemplate.h"
+#include "python-logtemplate-options.h"
+#include "python-logmsg.h"
+#include "python-helpers.h"
+#include "scratch-buffers.h"
+#include "messages.h"
+
+PyTypeObject py_log_template_type;
+
+void
+py_log_template_free(PyLogTemplate *self)
+{
+  log_template_unref(self->template);
+  g_free(self->template_options);
+  PyObject_Del(self);
+}
+
+PyObject *
+py_log_template_format(PyObject *s, PyObject *args, PyObject *kwrds)
+{
+  PyLogTemplate *self = (PyLogTemplate *)s;
+
+  PyLogMessage *msg;
+  PyLogTemplateOptions *log_template_options = NULL;
+  gint tz = LTZ_SEND;
+  gint seqnum = 0;
+
+  static const gchar *kwlist[] = {"msg", "template_options", "tz", "seqnum", NULL};
+
+  if (!PyArg_ParseTupleAndKeywords(args, kwrds, "O|Oii", (gchar **)kwlist,
+                                   &msg, &log_template_options, &tz, &seqnum))
+    return NULL;
+
+  if (Py_TYPE(msg) != &py_log_message_type)
+    {
+      PyErr_Format(PyExc_TypeError,
+                   "LogMessage expected in the first parameter");
+      return NULL;
+    }
+
+  if (log_template_options && (Py_TYPE(log_template_options) != &py_log_template_options_type))
+    {
+      PyErr_Format(PyExc_TypeError,
+                   "LogTemplateOptions expected in the second parameter");
+      return NULL;
+    }
+
+  LogTemplateOptions *template_options = log_template_options ? log_template_options->template_options :
+                                         self->template_options;
+  if (!template_options)
+    {
+      PyErr_Format(PyExc_RuntimeError,
+                   "LogTemplateOptions must be provided either in the LogTemplate constructor or as parameter of format");
+      return NULL;
+    }
+
+  GString *result = scratch_buffers_alloc();
+  log_template_format(self->template, msg->msg, template_options, tz, seqnum, NULL, result);
+
+  return _py_string_from_string(result->str, result->len);
+}
+
+PyObject *
+py_log_template_new(PyTypeObject *type, PyObject *args, PyObject *kwds)
+{
+  gchar *template_string;
+  PyLogTemplateOptions *log_template_options = NULL;
+  if (!PyArg_ParseTuple(args, "s|O", &template_string, &log_template_options))
+    return NULL;
+
+  if (log_template_options && (Py_TYPE(log_template_options) != &py_log_template_options_type))
+    {
+      PyErr_Format(PyExc_TypeError,
+                   "LogTemplateOptions expected in the second parameter");
+      return NULL;
+    }
+
+  LogTemplate *template = log_template_new(NULL, NULL);
+  GError *error = NULL;
+  if (!log_template_compile(template, template_string, &error))
+    {
+      PyErr_Format(PyExc_RuntimeError,
+                   "Error compiling template: %s", error->message);
```

Comment:
could we use a specific exception class? RuntimeError is very generic.

Reply:
Good idea! I created a new LogTemplateException class. Updated the description.